analyses/collapse-rnaseq/01-summarize_matrices.R

```diff
@@ -0,0 +1,103 @@
+
+suppressPackageStartupMessages(library(optparse))
+suppressPackageStartupMessages(library(tidyverse))
+suppressPackageStartupMessages(library(rtracklayer))
+
+option_list <- list(
+  make_option(c("-i", "--inputmat"), type = "character",
+              help = "Input matrix of merged RSEM files (.RDS)"),
+  make_option(c("-g", "--inputgtf"), type  = "character",
+              help = "Input gtf file (.gtf)"),
+  make_option(c("-m", "--outputmat"), type = "character",
+              help = "Output matrix (.RDS)"),
+  make_option(c("-t", "--dupstable"), type = "character",
+              help = "Output gene annotation table (.RDS)")
+)
+
+# parse the parameters
+opt <- parse_args(OptionParser(option_list = option_list))
+input.mat <- opt$inputmat
+input.gtf <- opt$inputgtf
+output.mat <- opt$outputmat
+dups.tab <- opt$dupstable
+
+if(file.exists(output.mat) & file.exists(dups.tab)) {
+  print("Matrix and Dups table exists!")
+} else {
+  print("Generating input matrix...!")
+  expr <- readRDS(input.mat)
+  
+  # read gencode (v27 for the paper) to get gene id-gene symbo-gene type annotation
+  # PAR_Y entries are just duplications - so remove beforehand
+  gtf <- rtracklayer::import(input.gtf)
+  gtf <- gtf %>% 
+    as.data.frame() %>%
+    select(gene_id, gene_name, gene_type) %>%
+    mutate(gene_id = str_replace_all(gene_id, "_PAR_Y", "")) %>%
+    unique()
+  
+  # split gene id and symbol
+  expr <- expr %>% 
+    mutate(gene_id = str_replace(gene_id, "_PAR_Y_", "_"))  %>%
+    separate(gene_id, c("gene_id", "gene_symbol"), sep = "\\_", extra = "merge") %>%
+    unique() 
+  
+  # remove all genes with no expression
+  expr <- expr[which(rowSums(expr[,3:ncol(expr)]) > 0),] 
+  gtf$expressed <- ifelse(gtf$gene_id %in% expr$gene_id, "Yes", "No")
+  
+  # identify gene symbols mapped to multiple ensembl identifiers
+  dups <- expr %>% 
+    select(gene_id, gene_symbol) %>%
+    group_by(gene_symbol) %>%
+    mutate(gene_symbol.ct = n()) %>%
+    filter(gene_symbol.ct > 1) %>%
+    unique()
+  gtf$ensembl_id <- ifelse(gtf$gene_id %in% dups$gene_id, "Multiple", "Unique")
+  
+  # collapse to matrix of HUGO symbols x Sample identifiers
+  # take mean per row and use the max value for duplicated gene symbols
+  expr.collapsed <- expr %>% 
+    mutate(means = rowMeans(select(.,-gene_id, -gene_symbol))) %>% # take rowMeans
+    arrange(desc(means)) %>% # arrange decreasing by means
+    distinct(gene_symbol, .keep_all = TRUE) %>% # keep the ones with greatest mean value. If ties occur, keep the first occurencce
+    select(-means) %>%
+    unique() %>%
+    remove_rownames() 
+  gtf$keep <- ifelse(gtf$gene_id %in% expr.collapsed$gene_id, "Yes", "No")
+  
+  # matrix of HUGO symbols x Sample identifiers
+  expr.input <- expr.collapsed %>% 
+    column_to_rownames("gene_symbol") %>%
+    select(-c(gene_id)) 
+  
+  print(dim(expr.input))
+  
+  # final geneid-symbol-biotype annotation file
+  print("Generating duplicates table...")
```

Comment:
This table isn't just duplicates, is it? In some places you call it the gene annotation table, which seems a better name. 

I don't think it matters much, but I did wonder if you might want to save this as a TSV for easier inspection outside R, since it doesn't seem that this has anything but text in it. Especially since you went to the trouble of converting the boolean values to "Yes/No".

Reply:
The table isnt just duplicates but it tells you what gene ids were kept, removed, which have duplicates gene symbols etc. I just went with the RDS convention that I have been seeing in the data files.